src/mpid/ch4/shm/posix/release_gather/release_gather.h

```diff
@@ -0,0 +1,337 @@
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+#ifndef RELEASE_GATHER_H_INCLUDED
+#define RELEASE_GATHER_H_INCLUDED
+
+extern zm_atomic_uint_t *MPIDI_POSIX_shm_limit_counter;
+extern MPL_shm_hnd_t shm_limit_handle;
+extern MPIDI_POSIX_release_gather_tree_type_t MPIDI_POSIX_Bcast_tree_type,
+    MPIDI_POSIX_Reduce_tree_type;
+
+/*Blocking wait implementation*/
+/* zm_memord_acquire makes sure no writes/reads are reordered before this load */
+#define WAIT_WHILE_LESS_THAN(ptr, value)                           \
+    do {                                                           \
+        int spin_count = 0;                                        \
+        while (zm_atomic_load(ptr, zm_memord_acquire) < (value)) { \
+            if (++spin_count >= 10000) {                           \
+                /* Call progress only after waiting for a while */ \
+                MPID_Progress_test();                              \
+                spin_count = 0;                                    \
+            }                                                      \
+        }                                                          \
+    }                                                              \
+    while (0)
+
+#define MPIDI_POSIX_RELEASE_GATHER_BCAST_CELLSIZE \
+    (MPIR_CVAR_BCAST_INTRANODE_BUFFER_TOTAL_SIZE / MPIR_CVAR_BCAST_INTRANODE_NUM_CELLS)
+#define MPIDI_POSIX_RELEASE_GATHER_REDUCE_CELLSIZE \
+    (MPIR_CVAR_REDUCE_INTRANODE_BUFFER_TOTAL_SIZE / MPIR_CVAR_REDUCE_INTRANODE_NUM_CELLS)
+#define FLAG_SIZE (sizeof(zm_atomic_uint_t))
+#define FLAG_SPACE_PER_RANK (MPIDU_SHM_CACHE_LINE_LEN * 2)      /* 1 cache_line each for gather and release flag */
+#define GATHER_FLAG_OFFSET (0)
+#define RELEASE_FLAG_OFFSET (MPIDU_SHM_CACHE_LINE_LEN)
+#define COMM_FIELD(comm, field)                   \
+    MPIDI_POSIX_COMM(comm)->release_gather->field
+
+#define GATHER_FLAG_ADDR(rank)                                    \
+    (((zm_atomic_uint_t *)release_gather_info_ptr->flags_addr) +  \
+    ((rank * FLAG_SPACE_PER_RANK + GATHER_FLAG_OFFSET)/(FLAG_SIZE)))
+#define RELEASE_FLAG_ADDR(rank)                                   \
+    (((zm_atomic_uint_t *)release_gather_info_ptr->flags_addr) +  \
+    ((rank * FLAG_SPACE_PER_RANK + RELEASE_FLAG_OFFSET)/(FLAG_SIZE)))
+#define BCAST_DATA_ADDR(buf)                           \
+    (char *) release_gather_info_ptr->bcast_buf_addr + \
+    (buf * MPIDI_POSIX_RELEASE_GATHER_BCAST_CELLSIZE)
+#define REDUCE_DATA_ADDR(rank, buf)                         \
+    (((char *) release_gather_info_ptr->reduce_buf_addr) +  \
+    (rank * MPIR_CVAR_REDUCE_INTRANODE_BUFFER_TOTAL_SIZE) + \
+    (buf * MPIDI_POSIX_RELEASE_GATHER_REDUCE_CELLSIZE))
+
```

Comment:
The macros above from `FLAG_SIZE` to `REDUCE_DATA_ADDR` should have a prefix `MPIDI_POSIX_RELEASE_GATHER_`, if they need to be kept in the `.h` file. If a macro is only used in `release_gather.c`, then it should be moved to the `.c` file and it can be defined without a namespace prefix in there.